scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java

```diff
@@ -506,41 +537,47 @@ private void terminateConnection(Connection connection, AlertMessage alert, DTLS
 		connectionClosed(connection.getPeerAddress());
 	}
 
-	private void processApplicationDataRecord(InetSocketAddress peerAddress, Record record)
-			throws GeneralSecurityException, HandshakeException {
+	private void processApplicationDataRecord(Record record) {
 
-		Connection connection = connectionStore.get(peerAddress);
-		DTLSSession session = null;
-		if (connection != null)
-			session = connection.getEstablishedSession();
-
-		if (session != null) {
+		Connection connection = connectionStore.get(record.getPeerAddress());
+		if (connection != null && connection.hasEstablishedSession()) {
+			DTLSSession session = connection.getEstablishedSession();
 			synchronized (session) {
 				// The DTLS 1.2 spec (section 4.1.2.6) advises to do replay detection
 				// before MAC validation based on the record's sequence numbers
 				// see http://tools.ietf.org/html/rfc6347#section-4.1.2.6
 				if (session.isRecordProcessable(record.getEpoch(), record.getSequenceNumber())) {
-					// APPLICATION_DATA can only be processed within the context of
-					// an established, i.e. fully negotiated, session
-					record.setSession(session);
-					ApplicationMessage message = (ApplicationMessage) record.getFragment();
-					// the fragment could be de-crypted
-					// thus, the handshake seems to have been completed successfully
-					connection.handshakeCompleted(peerAddress);
-					session.markRecordAsRead(record.getEpoch(), record.getSequenceNumber());
-					// finally, forward de-crypted message to application layer
-					if (messageHandler != null) {
-						messageHandler.receiveData(new RawData(message.getData(), peerAddress, session.getPeerIdentity()));
+					try {
+						// APPLICATION_DATA can only be processed within the context of
+						// an established, i.e. fully negotiated, session
+						record.setSession(session);
+						ApplicationMessage message = (ApplicationMessage) record.getFragment();
+						// the fragment could be de-crypted
+						// thus, the handshake seems to have been completed successfully
+						connection.handshakeCompleted(record.getPeerAddress());
+						session.markRecordAsRead(record.getEpoch(), record.getSequenceNumber());
+						// finally, forward de-crypted message to application layer
+						if (messageHandler != null) {
+							messageHandler.receiveData(new RawData(message.getData(), record.getPeerAddress(), session.getPeerIdentity()));
+						}
+					} catch (HandshakeException e) {
+						throw getDtlsHandshakeException(e, record.getPeerAddress());
+					} catch (GeneralSecurityException e) {
+						throw new DtlsHandshakeException(
+								String.format("Cannot de-crypt %s message from peer", record.getType()),
```

Comment:
I don't think it's a matter of taste :trollface:  https://dictionary.cambridge.org/spellcheck/english/?q=de-crypt